File: game-server/src/main/java/org/coinen/reactive/pacman/service/support/DefaultExtrasService.java
package org.coinen.reactive.pacman.service.support;

import org.coinen.pacman.Extra;
import org.coinen.reactive.pacman.repository.ExtrasRepository;
import org.coinen.reactive.pacman.repository.PlayerRepository;
import org.coinen.reactive.pacman.service.ExtrasService;
import reactor.core.Disposable;
import reactor.core.publisher.DirectProcessor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.FluxSink;
import reactor.core.publisher.Mono;

import java.time.Duration;

public class DefaultExtrasService implements ExtrasService {
    final ExtrasRepository extrasRepository;
    final PlayerRepository playerRepository;
    final DirectProcessor<Extra> extrasProcessor = DirectProcessor.create();
    final FluxSink<Extra> extrasFluxSink = extrasProcessor.serialize().sink();

    volatile Disposable powerUpTimer;
    volatile boolean powerUpActive;

    public DefaultExtrasService(ExtrasRepository extrasRepository,
        PlayerRepository playerRepository) {
        this.extrasRepository = extrasRepository;
        this.playerRepository = playerRepository;

        extrasRepository.saveAll(generate(60, 60, 11));
    }

    @Override
    public Flux<Extra> extras() {
        return extrasProcessor;
    }

    @Override
    public int check(float x, float y) {
        var retainedExtra = extrasRepository.collideExtra(x, y);

        if (retainedExtra != 0) {
            if (Math.signum(retainedExtra) == -1.0f) {
                if (this.powerUpActive)
                    this.powerUpTimer.dispose();
                this.powerUpTimer = Mono.delay(Duration.ofMillis(10000))
                    .doOnNext(e -> setPowerup(false))
                    .subscribe();
                this.powerUpActive = true;
                // scoreProcessor.onNext({player, score: player.getScore() + 1});
            }
            else {
                var sec = 10;
//                                   store.setPowerUpEnd(Instant.now() + sec * 1000);
            }
            var addedExtra = extrasRepository.createExtra(playerRepository.findAll()
                                                                          .size());

            var extra = Extra.newBuilder()
                             .setLast(retainedExtra)
                             .setCurrent(addedExtra)
                             .build();

            extrasFluxSink.next(extra);

            return retainedExtra;
        }

        return 0;
    }

    @Override
    public boolean isPowerupActive() {
        return this.powerUpActive;
    }

    private void setPowerup(boolean value) {
        this.powerUpActive = value;
    }

    static int[] generate(int width, int height, int offset) {
        var iterations =
            (int) ((width - 2 * offset) * (height - 2 * offset) * (0.3 + Math.random() * 0.3));
        var extras = new int[iterations];

        for (var i = 0; i < iterations; i++) {
            extras[i] = randomPosition(width, height, offset);
        }

        return extras;
    }

    static int randomPosition(int width, int height, int offset) {
        return (int) (Math.floor(Math.random() * (width - 2 * offset + 1) + offset) +  Math.floor(Math.random() * (height - 2 * offset + 1) + offset) * height);
    }
}

File: game-server/src/main/typescript/service/support/DefaultExtrasService.ts
import {DirectProcessor, Flux} from 'reactor-core-js/flux';
import {Extra} from 'game-idl';
import ExtrasService from '../ExtrasService';
import {ExtrasRepository, PlayerRepository} from '../../repository';

export default class DefaultExtrasService implements ExtrasService {
    extrasProcessor: DirectProcessor<Extra> = new DirectProcessor();

    powerUpTimer: NodeJS.Timeout;
    powerUpActive: boolean;

    constructor(
        private extrasRepository: ExtrasRepository,
        private playerRepository: PlayerRepository
    ) {
        extrasRepository.saveAll(DefaultExtrasService.generate(60, 60, 11))
    }


    extras(): Flux<Extra> {
        return this.extrasProcessor;
    }

    check(x: number, y: number): number {
        const retainedExtra = this.extrasRepository.collideExtra(x, y);
        
        if (retainedExtra != 0) {
            if (Math.sign(retainedExtra) < 0) {
                if (this.powerUpActive)
                    clearTimeout(this.powerUpTimer);
                
                
                this.powerUpTimer = setTimeout(this.setPowerup.bind(this, false), 10000);
                this.powerUpActive = true;
            }
            const addedExtra = this.extrasRepository
                .createExtra(this.playerRepository.findAll().length);

            const extra = new Extra();
            extra.setLast(retainedExtra);
            extra.setCurrent(addedExtra); 
            this.extrasProcessor.onNext(extra);

            return retainedExtra;
        }

        return 0;
    }

    public isPowerupActive(): boolean {
        return this.powerUpActive;
    }

    private setPowerup(value: boolean): void {
        this.powerUpActive = value;
    }

    static generate(width: number, height: number, offset: number): number[] {
        console.log('generating food');
        const iterations =
            ((width - 2 * offset) * (height - 2 * offset) * (0.3 + Math.random() * 0.3));
        const extras = [];
        console.log('iterations', iterations);
        for (let i = 0; i < iterations; i++) {
            extras[i] = DefaultExtrasService.randomPosition(width, height, offset);
        }
        console.log('generated', extras);
        return extras;
    }

    static randomPosition(width: number, height: number, offset: number): number {
        return Math.round(Math.floor(Math.random() * (width - 2 * offset + 1) + offset) +  Math.floor(Math.random() * (height - 2 * offset + 1) + offset) * height);
    }
}
